### PR TITLE
Create unique scopes for extensions

### PR DESCRIPTION
**Pull request by narrieta**
---

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made and Travis.CI is passing.

### Quality of Code and Contribution Guidelines
- [ ] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

---
This change is [Reviewable](https://reviewable.io/reviews/azure/walinuxagent/1549)

### CODE REVIEW

File: azurelinuxagent/common/cgroupapi.py [hunk 1]
```diff
@@ -278,18 +279,23 @@ def start_extension_command(self, extension_name, command, shell, cwd, env, stdo
         :param stdout: File object to redirect stdout to
         :param stderr: File object to redirect stderr to
         """
+        try:
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupapi.py]: I also fixed the issue where the cgroups directories were being deleted... now I re-create them on each command (if they do not exist)

File: azurelinuxagent/common/cgroupapi.py [hunk 2]
```diff
@@ -391,24 +394,38 @@ def remove_extension_cgroups(self, extension_name):
...
-        return [CGroup.create(cpu_cgroup_path, 'cpu', extension_name),
-                CGroup.create(memory_cgroup_path, 'cpu', extension_name)]
+        self._foreach_controller(create_cgroup, 'Cannot retrieve cgroup for extension {0}; resource usage will not be tracked.'.format(extension_name))
+
+        return cgroups
 
     def start_extension_command(self, extension_name, command, shell, cwd, env, stdout, stderr):
-        def pre_exec_function():
-            os.setsid()
+        scope_name = "{0}.{1}".format(self._get_extension_cgroup_name(extension_name), uuid.uuid4())
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupapi.py]: will use "{0}_{1}" per @vrdmr request

File: azurelinuxagent/common/cgroupapi.py [hunk 3]
```diff
@@ -391,24 +394,38 @@ def remove_extension_cgroups(self, extension_name):
...
     def get_extension_cgroups(self, extension_name):
-        scope_name = self._get_extension_scope_name(extension_name)
-        cpu_cgroup_path = os.path.join(CGROUPS_FILE_SYSTEM_ROOT, 'cpu', 'system.slice', scope_name)
-        memory_cgroup_path = os.path.join(CGROUPS_FILE_SYSTEM_ROOT, 'memory', 'system.slice', scope_name)
+        slice_name = self._get_extension_cgroup_name(extension_name)
+
+        cgroups = []
+
+        def create_cgroup(controller):
+            cpu_cgroup_path = os.path.join(CGROUPS_FILE_SYSTEM_ROOT, controller, 'system.slice', slice_name)
```
Comment (vrdmr) [inline azurelinuxagent/common/cgroupapi.py]: Just checking - `.scope` goes in here. Correct?

File: azurelinuxagent/common/cgroupapi.py [hunk 4]
```diff
@@ -109,6 +110,19 @@ def _get_controller_id(controller):
...
+        """
+        Executes the given operation on all controllers that need to be tracked; outputs 'message' if the controller is not mounted.
+        """
+        mounted_controllers = os.listdir(CGROUPS_FILE_SYSTEM_ROOT)
+
+        for controller in CGROUP_CONTROLLERS:
+            if controller not in mounted_controllers:
+                logger.warn('Controller "{0}" is not mounted. {1}'.format(controller, message))
+            else:
+                operation(controller)
```
[the remark this answers is not in the document]
Comment (narrieta) [reply azurelinuxagent/common/cgroupapi.py]: Yes, will be a separate PR

File: azurelinuxagent/common/cgroupapi.py [hunk 5]
```diff
@@ -302,7 +308,7 @@ def add_pid(controller):
             env=env,
             preexec_fn=pre_exec_function)
 
-        return process
+        return process, extension_cgroups
```
[the remark this answers is not in the document]
Comment (narrieta) [reply azurelinuxagent/common/cgroupapi.py]: we need a better way to send those cgroups to the telemetry thread. we talked about this issue and also the current race condition clearing tracked. let's follow up offline.

File: azurelinuxagent/common/cgroupapi.py [hunk 6]
```diff
@@ -370,9 +376,6 @@ def create_extension_cgroups(self, extension_name):
         self.create_and_start_unit(unit_filename, unit_contents)
         logger.info("Created slice for {0}".format(unit_filename))
```
[the remark this answers is not in the document]
Comment (narrieta) [reply azurelinuxagent/common/cgroupapi.py]: The instance you point out is in the implementation of the file system api; in that case create_extension_cgroups does return the correct values.
I will change the implementation of create_extension_cgroups in the systemd api to return an empty list. This function will be useful only after we create the extension scopes within the slice for the extension.
For now I will remote the usage of the return value from exthandlerinstance.

File: azurelinuxagent/common/cgroupapi.py (hunk 3 above)
Comment (narrieta) [reply azurelinuxagent/common/cgroupapi.py]: This one actually returns the slice for the extension (which we are not using currently). I'll add ".scope" in start_extension_command

**Comment by narrieta** [comment]
@vrdmr - Thanks for the review! I replied to your questions and did a few updates. Can you take a look at the last commit?

**Pull request merged**